onnxscript/rewriter/ort_fusions/attention.py

```diff
@@ -0,0 +1,188 @@
+        qkv_bias,
+        # mask_index,
+        past,
+        attention_bias,
+        num_heads,
+        scale,
+    ):
+        projected = op.MatMul(input, qkv_weight)
+        # Add bias if present
+        if self._has_input_bias:
+            projected = op.Add(projected, qkv_bias)
+
+        # Slice packed Matmul QKV into Q, K, and V
+        # Q, K, and V are of shape (B, S, D)
+        query_BSD = op.Slice(
+            projected,
+            _allow_other_inputs=True,
+            _allow_other_attributes=True,
+            _outputs=["query_mm_sliced"],
+        )
+        key_BSD = op.Slice(
+            projected,
+            _allow_other_inputs=True,
+            _allow_other_attributes=True,
+            _outputs=["key_mm_sliced"],
+        )
+        value_BSD = op.Slice(
+            projected,
+            _allow_other_inputs=True,
+            _allow_other_attributes=True,
+            _outputs=["value_mm_sliced"],
+        )
+
+        # Split past into past_key and past_value
+        # past_key and past_value are of shape (B, H, S, D/H)
+        past_key, past_value = op.Split(past, axis=0, split=[1, 1])
+
+        # TODO: Pass other attributes
+        attention, present_key, present_value = op.MultiHeadAttention(
+            query_BSD,
+            key_BSD,
+            value_BSD,
+            None,  # bias
+            None,  # key_padding_mask
+            attention_bias,
+            past_key,
+            past_value,
+            num_heads=num_heads,
+            scale=scale,
+            _domain="com.microsoft",
+            _outputs=3,
+        )
+
+        # Concat present_key and present_value to form present
+        present = op.Concat(present_key, present_value, axis=0)
+        return attention, present
+
+    def check(
+        self,
+        op,
+        input,
+        qkv_weight,
+        qkv_bias,
+        query_mm_sliced,
+        key_mm_sliced,
+        value_mm_sliced,
+        **_,
+    ):
+        bindings: dict[str, Dim] = {}
+
+        def no_match(val: ir.Value, dims: Sequence[str]) -> bool:
+            return not _check_shape(bindings, val, dims)
+
+        if no_match(input, ["B", "S", "D"]):
+            return False
+        if no_match(qkv_weight, ["D", "Dh"]):
+            return False
+        if no_match(qkv_bias, ["Dh"]):
+            return False
+        if no_match(query_mm_sliced, ["B", "S", "Dh_q"]):
+            return False
+        if no_match(key_mm_sliced, ["B", "S", "Dh_k"]):
+            return False
+        if no_match(value_mm_sliced, ["B", "S", "Dh_v"]):
+            return False
+
+        # Ensure Dh = Dh_q + Dh_k + Dh_v
+        Dh = bindings.get("Dh")
+        Dh_q = bindings.get("Dh_q")
+        Dh_k = bindings.get("Dh_k")
+        Dh_v = bindings.get("Dh_v")
+
+        if Dh is None or Dh_q is None or Dh_k is None or Dh_v is None:
+            return False  # Missing bindings, cannot verify
+
+        if Dh != Dh_q + Dh_k + Dh_v:  # type: ignore[operator]
```

Comment:
The dimensions can be symbolic (in worst case) ... which is why there is operator warning. May be worth changing previous check from `Dh is None` to `not instance(Dh, int)`.